core/src/main/java/hudson/model/DirectoryBrowserSupport.java

```diff
@@ -436,33 +437,81 @@ private static void zip(StaplerResponse rsp, VirtualFile root, VirtualFile dir,
         try (ZipOutputStream zos = new ZipOutputStream(outputStream)) {
             zos.setEncoding(System.getProperty("file.encoding")); // TODO JENKINS-20663 make this overridable via query parameter
             // TODO consider using run(Callable) here
```

Comment:
Is this todo still applicable?

Reply:
@jglick Do you remember why you put the TODO there ? 

The initial code used the `.list(...)` method that was already a callable, but the stream comes from the response. Did you think about sending the zip from the callable directly ?

Reply:
No, `run(Callable)` is about batch computation. Not to be confused with `hudson.remoting.Callable`!

See [JEP-202](https://jenkins.io/jep/202). Currently when you download a `*.zip` URL containing a number of artifacts (for example the Incrementals publisher function does this), some `VirtualFile` methods (`list`, `open`, `lastModified`) will be called one by one, asymptotically proportional to the number of files. For artifacts stored in `$JENKINS_HOME`, this is fine, but if you are using `artifact-manager-s3` it would be rather inefficient. Using `VirtualFile.run(Callable)` here would allow for a faster implementation—in the case of S3, one metadata API call plus one blob download per artifact.

(Even better would be a URL pattern returning some text response like

```
target/stuff-1.0.pom https://s3.whatever…/…/target/stuff-1.0.pom?authentication=…
target/stuff-1.0.jar https://s3.whatever…/…/target/stuff-1.0.jar?authentication=…
```

so that the master could avoid downloading blobs at all.)

Reply:
@jglick thanks for the precision, really appreciated. Are you OK if we let the TODO in place for future improvement ?

Reply:
If you are not intending to introduce a `run(Callable)` call then the comment should be left alone.